### PR TITLE
feat: introduction of tooltips to trigger-buttons in AppLayout VisualRefresh

### PR DESCRIPTION
**Pull request by dpitcock**
### Description

- Adds logic to render tooltips for the trigger-buttons in toolbar
- Pulls the trigger-buttons for mobile and desktop into the same file with a `isMobile` prop that renders accordingly yet the tooltip logic can live in its wrapping element
- when using the trigger-button, it by default has no tooltip, with when the props `hasTooltip` it will then render a tooltip if there is another prop for `tooltipText `or `ariaLabel` as a fallback. This way the trigger-button won't create a tooltip for other uses
- when the trigger-button is rendered without tooltip, there are no eventlisteners being created for it.

Will update with more with next PR

Related links, issue #, if available: n/a

### How has this been tested?

- Unit tests created for the Trigger-button explicitly
- Assertions of tooltips are also added in the drawers tests

<details>
   <summary>Review checklist</summary>

You can see the UI changes on this [page deployment](https://d21d5uik3ws71m.cloudfront.net/components/3c6894e76f768313eedc7e5de30050ea278b5780/dev-pages/index.html#/light/app-layout/with-drawers)  Note on both mobile and desktop:

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_
- _Changes are covered with new/existing integration tests?_
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/app-layout/__tests__/drawers.test.tsx [hunk 1]
```diff
@@ -169,4 +173,38 @@ describeEachAppLayout(({ size }) => {
...
+
+    const otherEl = wrapper!.findNavigation();
+    expect(otherEl).toBeTruthy();
+    otherEl.fireEvent(new MouseEvent('pointerdown', { bubbles: true }));
+    waitFor(() => {
+      expect(wrapper!.findByClassName(tooltipStyles.root)).toBeNull();
+    });
+  });
+
+  testIf(size !== 'mobile')('tooltip renders correctly on focus and blue events', () => {
```
Comment (dpitcock) [inline src/app-layout/__tests__/drawers.test.tsx:195]: I will update blue === blur

File: src/app-layout/__tests__/drawers.test.tsx [hunk 2]
```diff
@@ -169,4 +173,38 @@ describeEachAppLayout(({ size }) => {
...
+    const drawerTrigger = wrapper.findDrawerTriggerById('security')!;
+
+    drawerTrigger!.focus();
+    waitFor(() => {
+      expect(wrapper!.findByClassName(tooltipStyles.root)).toBeTruthy();
+      expect(wrapper!.findByClassName(popoverStyles.content)).toBeTruthy();
+    });
+
+    drawerTrigger!.blur();
+    waitFor(() => {
```
Comment (dpitcock) [inline src/app-layout/__tests__/drawers.test.tsx:206]: I will also confirm if the waitFor is neccessary and will  await for it if it is and remove if not

File: src/app-layout/__tests__/trigger-button.test.tsx [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+      },
+      ref as any
+    );
+    expect(wrapper).not.toBeNull();
+    const button = wrapper.find('button')!;
+    button.click();
+    expect(mockClickSpy).toBeCalledTimes(disabledValue ? 0 : 1);
+  });
+
+  it.skip('does not render without either an iconName or iconSVG prop', async () => {
```
Comment (dpitcock) [inline src/app-layout/__tests__/trigger-button.test.tsx:108]: Will discuss what should happen if no iconName nor iconSvg provided.

File: src/app-layout/__tests__/trigger-button.test.tsx [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+function delay() {
+  return act(() => new Promise(resolve => setTimeout(resolve)));
+}
+
+async function renderTriggerButton(props: Partial<TriggerButtonProps> = {}, ref: React.Ref<ButtonProps.Ref>) {
+  const renderProps = { ...mockProps, ...props };
+  const { container, rerender, getByTestId, getByText } = render(
+    <div>
+      <TriggerButton {...(renderProps as TriggerButtonProps)} ref={ref} />
+      <span className={mockOtherEl.class}>{mockOtherEl.text}</span>
```
Comment (dpitcock) [inline src/app-layout/__tests__/trigger-button.test.tsx:37]: Note to self, I need to add one more test that clicks on the outer element here to confirm tooltip closes

File: src/app-layout/visual-refresh/drawers.tsx [hunk 5]
```diff
@@ -238,6 +238,7 @@ function DesktopTriggers() {
           return (
             <TriggerButton
               ariaLabel={item.ariaLabels?.triggerButton}
+              tooltipText={item.ariaLabels?.drawerName}
```
Comment (dpitcock) [inline src/app-layout/visual-refresh/drawers.tsx:241]: Note to self, a new test should assert the Aria Label works as the default if no tooltipText given

**Review by dpitcock — COMMENTED** [review]
Notes to self for next revision

File: src/app-layout/__tests__/drawers.test.tsx (hunk 1 above)
Comment (dpitcock) [reply src/app-layout/__tests__/drawers.test.tsx:195]: I will also expand to mobile and see if the waitFor is needed or needs to be awaited.

File: src/app-layout/visual-refresh/trigger-button.tsx [hunk 6]
```diff
@@ -1,23 +1,53 @@
...
+   *
+   * When falsy, the tooltip will parse the tooltip form the aria-lable
+   */
+  tooltipText?: string;
+  /**
+   * Ovewrwrites any internal testIds when provided
+   */
   testId?: string;
-  onClick: React.MouseEventHandler<HTMLButtonElement>;
+  onClick: React.MouseEventHandler<HTMLButtonElement> | CancelableEventHandler<ClickDetail>;
```
Comment (just-boris) [inline src/app-layout/visual-refresh/trigger-button.tsx:50]:
```suggestion
  onClick: ButtonProps['onClick']
```

File: src/app-layout/visual-refresh/trigger-button.tsx [hunk 7]
```diff
@@ -1,24 +1,61 @@
...
+   */
+  isMobile?: boolean;
   ariaControls?: string;
+  disabled?: boolean;
+  /**
+   * If the button is expected to have a tooltip. When false it will not set the event listeners
+   *
+   * defaults to false
+   */
+  hasTooltip?: boolean;
```
Comment (dpitcock) [inline src/app-layout/visual-refresh/trigger-button.tsx:43]: Added this prop so trigger-buttons that do not have tooltips do not create event listeners

File: src/app-layout/visual-refresh/trigger-button.tsx [hunk 8]
```diff
@@ -32,38 +69,134 @@ function TriggerButton(
...
+            onClick={onClick}
+            ref={ref as Ref<HTMLButtonElement>}
+            type="button"
+            data-testid={testId}
+          >
+            <span className={clsx(badge && styles['trigger-badge-wrapper'])}>
+              {(iconName || iconSvg) && <Icon name={iconName} svg={iconSvg} />}
+            </span>
+          </button>
+          {badge && <div className={styles.dot} />}
```
Comment (dpitcock) [inline src/app-layout/visual-refresh/trigger-button.tsx:189]: InternalButton has it's own badge rendering so a badge only needs to be added for desktop

**Review by dpitcock — COMMENTED** [review]
Notes on latest commit

File: src/app-layout/visual-refresh/trigger-button.tsx [hunk 9]
```diff
@@ -32,38 +69,134 @@ function TriggerButton(
...
-        </span>
-      </button>
-      {badge && <div className={styles.dot} />}
+    <div
+      ref={containerRef}
+      {...(hasTooltip
+        ? {
+            onPointerEnter: () => onShowTooltipSoft(true),
+            onPointerLeave: () => onShowTooltipSoft(false),
+            onFocus: ({ target, currentTarget }) => {
```
Comment (dpitcock) [inline src/app-layout/visual-refresh/trigger-button.tsx:131]: I'm looking into creating a custom focus handler that will check for details in the event, and not show the tooltip is a detail stating it was triggered from a drawer close via the close button on the drawer

**Review by dpitcock — COMMENTED** [review]
Accessibility solution.  Keep tooltip showing on keyboard navigation yet disable tooltip when focused from drawer close

File: src/app-layout/__integ__/app-layout-toolbar-tooltips-keys.test.ts [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+      await page.pause(100);
+      await expect(page.isExisting(`.${visualRefreshStyles['trigger-tooltip']}`)).resolves.toBeFalsy();
+      await expect(
+        page.isExisting(`.${visualRefreshStyles[`drawers-desktop-triggers-container`]}`)
+      ).resolves.toBeTruthy();
+      await expect(
+        page.isExisting(`.${visualRefreshStyles['drawers-desktop-triggers-container']}`)
+      ).resolves.toBeTruthy();
+
+      drawerIds.map(async (drawerId: string) => {
```
Comment (just-boris) [inline src/app-layout/__integ__/app-layout-toolbar-tooltips-keys.test.ts:61]:
```suggestion
      for(const drawerId of drawerIds) {
```

**Pull request closed without merging**